eth_p2p/rlpx_protocols/shh_protocol.nim

```diff
@@ -416,31 +539,406 @@ proc add*(self: var Queue, msg: Message) =
     discard
 
   proc messages(peer: Peer, envelopes: openarray[Envelope]) =
-    discard
+    if not peer.state.initialized:
+      warn "Handshake not completed yet, discarding messages"
+      return
 
-  proc powRequirement(peer: Peer, value: float64) =
-    discard
+    for envelope in envelopes:
+      # check if expired or in future, or ttl not 0
+      if not envelope.valid():
+        warn "Expired or future timed envelope"
+        # disconnect from peers sending bad envelopes
+        # await peer.disconnect(SubprotocolReason)
+        continue
+
+      let msg = initMessage(envelope)
+      if not msg.allowed(peer.networkState.config):
+        # disconnect from peers sending bad envelopes
+        # await peer.disconnect(SubprotocolReason)
+        continue
+
+      # This peer send it thus should not receive it again
+      peer.state(shh).received.incl(msg)
+
+      if peer.networkState.queue.add(msg):
+        # notify filters of this message
+        peer.networkState.filters.notify(msg)
+
+  proc powRequirement(peer: Peer, value: uint) =
+    if not peer.state.initialized:
+      warn "Handshake not completed yet, discarding powRequirement"
+      return
+
+    peer.state.powRequirement = cast[float64](value)
 
   proc bloomFilterExchange(peer: Peer, bloom: Bytes) =
-    discard
+    if not peer.state.initialized:
+      warn "Handshake not completed yet, discarding bloomFilterExchange"
+      return
+
+    peer.state.bloom.bytesCopy(bloom)
 
   nextID 126
 
   proc p2pRequest(peer: Peer, envelope: Envelope) =
+    # TODO: here we would have to allow to insert some specific implementation
+    # such as e.g. Whisper Mail Server
     discard
 
   proc p2pMessage(peer: Peer, envelope: Envelope) =
-    discard
+    if peer.state.trusted:
+      # when trusted we can bypass any checks on envelope
+      let msg = Message(env: envelope, isP2P: true)
+      peer.networkState.filters.notify(msg)
+
+# 'Runner' calls ---------------------------------------------------------------
+
+proc processQueue(peer: Peer) =
+  var envelopes: seq[Envelope] = @[]
+  for message in peer.networkState(shh).queue.items:
+    if peer.state(shh).received.contains(message):
+      # debug "message was already send to peer"
+      continue
+
+    if message.pow < peer.state(shh).powRequirement:
+      debug "Message PoW too low for peer"
+      continue
+
+    if not bloomFilterMatch(peer.state(shh).bloom, message.bloom):
+      debug "Message does not match peer bloom filter"
+      continue
+
+    debug "Adding envelope"
+    envelopes.add(message.env)
+    peer.state(shh).received.incl(message)
+
+  debug "Sending envelopes", amount=envelopes.len
+  # await peer.messages(envelopes)
+  asyncCheck peer.messages(envelopes)
+
+proc run(peer: Peer) {.async.} =
+  peer.state(shh).running = true
+  while peer.state(shh).running:
+    if not peer.networkState(shh).config.isLightNode:
+      peer.processQueue()
+    await sleepAsync(300)
```

Comment:
Is this how the other clients do it? The polling interval seems to be quite high and I'm wondering whether it's possible to use a more event-driven way to handle everything.

Reply:
Well, I only checked geth, and that is how it is done there, with those intervals. The polling interval on the `processQueue` can be lowered, of course, finding the balance to remain a good 'citizen' on the network.

Initially I had in mind of doing it more event-driven, e.g. forwarding messages on the event of receiving them (it does make things a little more complex, like processing the queue on joining peers, peers that change bloomfilter, PoW, etc.). Is this what you mean for example? 

It is not in the spec (EIP-627), however it is mentioned here that messages should be added and then processed in bulk: https://github.com/ethereum/go-ethereum/wiki/Whisper-Overview#creating-and-sending-messages 

While it is not mentioned there why, I assumed that it is done most likely to maintain "darkness". Imagine the implementation always forwards messages directly when they are being received. Then if your own "posted" message is send in solo the peers can probably know it originates from you more easily by doing some network statistics.

After some very quick thought some example ideas:
1. Some network monitoring could show that you posted a message without initially receiving any other messages? (perhaps invalid due to high amount of traffic)
2. Most other peers send (bigger) lists of envelopes (as it is implemented like this now to stack them up for 300 ms in geth), thus you will also forward these lists of envelopes. Your single message post will be spotted easily. (I guess this does not hold if every implementation does it more event driven)

These are just some very quick thoughts and I'm not sure how valid they are but I guess what I am saying is that it sure feels that this way of queuing the messages up is hiding the own posted messages better.

Perhaps more complex tricks could be done like sending message at random time after receiving and such, but I am not sure if this will benefit the solution in the end.

Reply:
OK, thanks for the detailed explanation. I guess copying Geth is the safest route right now.